pkg/mtr/mtr.go

```diff
@@ -161,37 +172,47 @@ func (m *MTR) discover(count int) {
 	ipAddr := net.IPAddr{IP: net.ParseIP(m.Address)}
 
 	for i := 1; i <= count; i++ {
-		time.Sleep(m.interval)
-
-		unknownHopsCount := 0
-		for ttl := 1; ttl < m.maxHops; ttl++ {
-			seq++
-			time.Sleep(m.hopsleep)
-			var hopReturn icmp.ICMPReturn
-			var err error
-			if ipAddr.IP.To4() != nil {
-				hopReturn, err = icmp.SendDiscoverICMP(m.SrcAddress, &ipAddr, ttl, id, m.timeout, seq)
-			} else {
-				hopReturn, err = icmp.SendDiscoverICMPv6(m.SrcAddress, &ipAddr, ttl, id, m.timeout, seq)
-			}
-
-			m.mutex.Lock()
-			s := m.registerStatistic(ttl, hopReturn)
-			s.Dest = &ipAddr
-			s.PID = id
-			m.mutex.Unlock()
-			m.channel <- struct{}{}
-			if hopReturn.Addr == m.Address {
-				break
-			}
-			if err != nil || !hopReturn.Success {
-				unknownHopsCount++
-				if unknownHopsCount >= m.maxUnknownHops {
-					break
+		select {
+		case <-ctx.Done():
+			return ErrTimeout
+		case <-time.After(m.interval):
+			unknownHopsCount := 0
+
+		ttlLoop:
```

Comment:
I'm not sure what the opinions are on label loops. We need to break the loop from inside the select, and this seemed the most convenient way to do it, but I'm not sure if that's considered a code smell or not.